services/localfile/server/localfile.go

```
@@ -132,89 +153,153 @@ func (s *server) Read(req *pb.ReadActionRequest, stream pb.LocalFile_ReadServer)
 	}
 
-	logger.Info("read request", "filename", file)
-	if err := util.ValidPath(file); err != nil {
-		recorder.CounterOrLog(ctx, localfileReadFailureCounter, 1, attribute.String("reason", "invalid_path"))
-		return err
-	}
-	f, err := os.Open(file)
-	if err != nil {
-		recorder.CounterOrLog(ctx, localfileReadFailureCounter, 1, attribute.String("reason", "open_err"))
-		return status.Errorf(codes.Internal, "can't open file %s: %v", file, err)
-	}
-
-	defer func() {
-		if err := f.Close(); err != nil {
-			recorder.CounterOrLog(ctx, localfileReadFailureCounter, 1, attribute.String("reason", "close_err"))
-			logger.Error(err, "file.Close()", "file", file)
-		}
-	}()
-
-	// Seek forward if requested
-	if offset != 0 {
-		whence := 0
-		// If negative we're tailing from the end so
-		// negate the sign and set whence.
-		if offset < 0 {
-			whence = 2
-		}
-		if _, err := f.Seek(offset, whence); err != nil {
-			recorder.CounterOrLog(ctx, localfileReadFailureCounter, 1, attribute.String("reason", "seek_err"))
-			return status.Errorf(codes.Internal, "can't seek for file %s: %v", file, err)
-		}
-	}
-
-	max := length
-	if max == 0 {
-		max = math.MaxInt64
-	}
-
-	buf := make([]byte, util.StreamingChunkSize)
-
-	reader := io.LimitReader(f, max)
-
-	td, closer, err := dataPrep(f)
-	if err != nil {
-		recorder.CounterOrLog(ctx, localfileReadFailureCounter, 1, attribute.String("reason", "dataprep_err"))
-		return err
-	}
-	defer closer()
-
-	for {
-		n, err := reader.Read(buf)
-		// If we got EOF we're done for normal reads and wait for tails.
-		if err == io.EOF {
-			// If we're not tailing then we're done.
-			if r != nil {
-				break
-			}
-			if err := dataReady(td, stream); err != nil {
-				recorder.CounterOrLog(ctx, localfileReadFailureCounter, 1, attribute.String("reason", "dataready_err"))
+	errs, ctx := errgroup.WithContext(ctx)
+	fileChan := make(chan string, 100)
+	if path, found := strings.CutSuffix(file, "/*"); found {
```

Any reason you didn't use `filepath.Glob`?

---

You could argue we get the List implementation done using `filepath.Glob` too, but since this codebase is already written there to be ready for recursive option and compatible with streaming of large number of files I wanted just to reuse List.

That being said I do not see clearly why list didn't got, at the fist cut, implemented with `filepath.Glob`, I do sense there might be some history there.

---

Possibly the consideration was that Glob eats errors:

`Glob ignores file system errors such as I/O errors reading directories. The only possible returned error is [ErrBadPattern](https://pkg.go.dev/path/filepath#ErrBadPattern), when pattern is malformed.`